Improve showing toric varieties

lkastner opened the pull request:
Refactored to be shorter, and logic untangled a little.

lkastner committed:
ToricVarieties: Untangle Base.show(NTV)

lkastner committed:
ToricVarieties: Add minimal logic back into Base.show(NTV)

lkastner committed:
ToricVarieties: First batch of docfixes

lkastner committed:
ToricVarieties: Only consider projective, if not affine

lkastner committed:
ToricVarieties: Reduce length of show(ToricDivisor) code

lkastner committed:
ToricVarieties: issimplicial needs to be cached, otherwise it will never be shown

lkastner committed:
ToricVarieties: Further condense and logically untangle show

lkastner committed:
ToricVarieties: Explain logic of output function

lkastner committed:
ToricVarieties: Fix bug in show method

lkastner committed:
ToricVarieties: Apparently it is is_prime_divisor

lkastner committed:
ToricVarieties: is_prime_divisor->isprime

lkastner committed:
ToricVarieties: Only display affine, as it almost implies non-complete

fingolfin commented:
@HereAround any comments? Approval?

fingolfin commented on src/ToricVarieties/NormalToricVarieties/constructors.jl:
```diff
@@ -479,95 +479,73 @@ function Base.:*(v::AbstractNormalToricVariety, w::AbstractNormalToricVariety)
+function push_attribute_if_exists!(result::Vector{String}, 
+        v::T,
+        property::Symbol, name::String, 
+        alt_string::Union{String,Nothing}=nothing; 
```
Could one also do this and then simplify the code accordingly? I can't try it myself right now, so it might be nonsense, but I thought this kind of code supported by Julia? Anyway, unimportant overall :-)
```suggestion
        alt_string::String="non-"*name; 
```

lkastner replied:
Thanks, that works and even removes a few lines. :)

lkastner committed:
ToricVarieties: Simplify default option

lkastner committed:
ToricVarieties: Gorenstein implies Q-Gorenstein

HereAround commented:
Thank you @lkastner. This looks good! In particular, I like the function `push_attribute_if_exists`, which simplifies `base.show` significantly. Well done!

Here are a few comments that I believe we should discuss:
1. I imagine that at some point in the future, someone might be looking at the display method for `ToricDivisors`, which uses `push_attribute_if_exists`, and then searches for some time until finding the implementation of that function `push_attribute_if_exists` in the constructors of `ToricVarieties`. Could we maybe add an `auxilliary.jl`-file at the top-level of `ToricVarieties` (i.e. next to the file `JToric.jl`) and implement the function in there?
2. Even more, once attribute caching is more frequently used in `Oscar` (e.g. in `base.show` methods), this function could be interesting to other people. Maybe it should be documented?
3. I believe also the display method for the `ToricLineBundles` should be updated (with the properties introduced in https://github.com/oscar-system/Oscar.jl/pull/941). It would seem reasonable to add this extension here. But of course, we can also do this in a separate PR if you prefer. In any case, this would again use `push_attribute_if_exists`, making it (from my perspective) again indicative to place that function at a central location.
4. How about the `cyclic_quotient_singularities`? Would we care to adjust/customize display methods towards them?
5. Is it really necessary to cache `IsSimplicial` of a toric variety? This is a synonym for `IsOrbifold`. So it would appear that this adds unnecessary memory for caching. In addition, due to https://github.com/oscar-system/Oscar.jl/issues/997, we would then probably triple-cache this property, by also storing the value in `Polymake`. I would suggest to revert to the previous implementation, which simply returns the value of `IsOrbifold`.

lkastner committed:
ToricVarieties: Add custom show method for cyclic quotient singularity

lkastner committed:
ToricVarieties: Move helper show method to more central location

lkastner committed:
ToricVarieties: issimplicial should access the same attribute as isorbifold

lkastner commented:
@HereAround 
1. For now I just moved the `push_attribute_if_exists` to the central file at the beginning, then it should be easy to find.
2. I recently learned about describe, etc., so I think we should definitely discuss this with the other developers, but right now I don't feel bold enough to put this code in a more central location. ;)
3. I don't know so much about toric line bundles, so I don't know all the implications at the top of my head. Hence I suggest a separate PR.
4. I changed the show here, but these are much more "rigid" than our other toric varieties, so it just prints the standard notation for these varieties.
5. I just introduced this, since otherwise it did not show up in the output anymore. Now I changed `issimplicial` to use the same attribute as `isorbifold`. This should be fine, right?

HereAround commented:
> @HereAround
> 
>     1. For now I just moved the `push_attribute_if_exists` to the central file at the beginning, then it should be easy to find.
> 
>     2. I recently learned about describe, etc., so I think we should definitely discuss this with the other developers, but right now I don't feel bold enough to put this code in a more central location. ;)
> 
>     3. I don't know so much about toric line bundles, so I don't know all the implications at the top of my head. Hence I suggest a separate PR.
> 
>     4. I changed the show here, but these are much more "rigid" than our other toric varieties, so it just prints the standard notation for these varieties.
> 
>     5. I just introduced this, since otherwise it did not show up in the output anymore. Now I changed `issimplicial` to use the same attribute as `isorbifold`. This should be fine, right?

@lkastner Sounds good! I approve once the tests go green.

lkastner committed:
ToricVarieties: Fix doctests with new Base.show(CQS)

HereAround reviewed (approved):
Looks great. Thank you @lkastner!
